lib/icinga/checkable-dependency.cpp

```diff
@@ -43,89 +184,72 @@ std::vector<Dependency::Ptr> Checkable::GetReverseDependencies() const
 		Host::Ptr host = service->GetHost();
 
 		if (host && host->GetState() != HostUp && host->GetStateType() == StateTypeHard) {
-			if (failedDependency)
-				*failedDependency = nullptr;
-
 			return false;
 		}
 	}
 
-	auto deps = GetDependencies();
-
-	std::unordered_map<std::string, Dependency::Ptr> violated; // key: redundancy group, value: nullptr if satisfied, violating dependency otherwise
-
-	for (const Dependency::Ptr& dep : deps) {
-		std::string redundancy_group = dep->GetRedundancyGroup();
+	for (auto& dependencyGroup : GetDependencyGroups()) {
+		if (auto state(dependencyGroup->GetState(this, dt, rstack + 1)); !state.Reachable || !state.OK) {
+			Log(LogDebug, "Checkable")
+				<< "Dependency group '" << dependencyGroup->GetRedundancyGroupName() << "' have failed for checkable '"
+				<< GetName() << "': Marking as unreachable.";
 
-		if (!dep->IsAvailable(dt)) {
-			if (redundancy_group.empty()) {
-				Log(LogDebug, "Checkable")
-					<< "Non-redundant dependency '" << dep->GetName() << "' failed for checkable '" << GetName() << "': Marking as unreachable.";
-
-				if (failedDependency)
-					*failedDependency = dep;
-
-				return false;
-			}
-
-			// tentatively mark this dependency group as failed unless it is already marked;
-			//  so it either passed before (don't overwrite) or already failed (so don't care)
-			// note that std::unordered_map::insert() will not overwrite an existing entry
-			violated.insert(std::make_pair(redundancy_group, dep));
-		} else if (!redundancy_group.empty()) {
-			violated[redundancy_group] = nullptr;
+			return false;
 		}
 	}
 
-	auto violator = std::find_if(violated.begin(), violated.end(), [](auto& v) { return v.second != nullptr; });
-	if (violator != violated.end()) {
-		Log(LogDebug, "Checkable")
-			<< "All dependencies in redundancy group '" << violator->first << "' have failed for checkable '" << GetName() << "': Marking as unreachable.";
-
-		if (failedDependency)
-			*failedDependency = violator->second;
+	return true;
+}
 
+/**
+ * Checks whether the last check result of this Checkable affects its child dependencies.
+ *
+ * A Checkable affects its child dependencies if it runs into a non-OK state and results in any of its child
+ * Checkables to become unreachable. Though, that unavailable dependency may not necessarily cause the child
+ * Checkable to be in unreachable state as it might have some other dependencies that are still reachable, instead
+ * it just indicates whether the edge/connection between this and the child Checkable is broken or not.
+ *
+ * @return bool - Returns true if the Checkable affects its child dependencies, otherwise false.
+ */
+bool Checkable::AffectsChildren(const CheckResult::Ptr& cr) const
+{
+	if (!cr || !IsReachable()) {
+		// If there is no check result, or the Checkable is not reachable, we can't safely determine whether
+		// the Checkable affects its child dependencies.
 		return false;
 	}
```

Comment:
As you've already noticed yourself, this doesn't really uses `cr` (besides checking it for `nullptr`), so that probably doesn't work as intended.